storage/rocksdb/rdb_locking_iter.h

```diff
@@ -0,0 +1,222 @@
+
+/* MySQL header files */
+#include "sql/handler.h"   /* handler */
+#include "sql/debug_sync.h"
+#include "./rdb_threads.h" /* for thd_get_current_thd */
+
+/* MyRocks header files */
+#include "./ha_rocksdb.h"
+
+namespace myrocks {
+
+//////////////////////////////////////////////////////////////////////////////
+// Locking iterator
+//////////////////////////////////////////////////////////////////////////////
+
+//
+// LockingIterator is an iterator that locks the rows before returning, as well
```

Comment:
I don't think this actually works with non-materialized partial indexes. In the case where partial index is not materialized, the rows are read from the primary key, sorted in the secondary key order and then returned directly by the partial_index iterator. It would not make Seek()/Next() calls on the secondary index down to rocksdb, and so the range locks in the non-materialized partial index case would not be acquired through this iterator.

I'm getting an assertion when trying to test this though:

```
create table t1 (a int, b int, c int, d int, e int default 0  primary key(a, b, c), key (a, b, d) comment 'partial_group_keyparts=2,partial_group_threshold=10000') engine=rocksdb;
insert into t1 (a,b,c,d) values (1, 1, 100, 1000), (1,1,200, 2000), (1, 1, 300, 3000), (1, 1, 400, 4000);
update t1 set d = d + 1 where a = 1 and b = 1 and d > 2000; 
```

```
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007ffff7e124ed in __GI_abort () at abort.c:79
#2  0x00007ffff7e224d3 in __assert_fail_base (fmt=0x7ffff7f9a088 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", 
    assertion=assertion@entry=0x3eda1fc "!key_descr.is_primary_key() || (key_descr.is_primary_key() && m_row_lock_count > 0)", 
    file=file@entry=0x3cdbdc3 "/home/herman/rocks-mysql/8.0/storage/rocksdb/ha_rocksdb.cc", line=line@entry=4040, 
    function=function@entry=0x3a0cb52 "virtual void myrocks::Rdb_transaction_impl::release_lock(const myrocks::Rdb_key_def &, const std::string &, bool)") at assert.c:92
#3  0x00007ffff7e22543 in __GI___assert_fail (
    assertion=0x3eda1fc "!key_descr.is_primary_key() || (key_descr.is_primary_key() && m_row_lock_count > 0)", 
    file=0x3cdbdc3 "/home/herman/rocks-mysql/8.0/storage/rocksdb/ha_rocksdb.cc", line=4040, 
    function=0x3a0cb52 "virtual void myrocks::Rdb_transaction_impl::release_lock(const myrocks::Rdb_key_def &, const std::string &, bool)")
    at assert.c:101
#4  0x0000000007b54dfb in myrocks::Rdb_transaction_impl::release_lock (this=0x7fffbc84aa80, key_descr=..., 
    rowkey="\000\000\001\004\200\000\000\001\200\000\000\001\200\000", <incomplete sequence \310>, force=false)
    at /home/herman/rocks-mysql/8.0/storage/rocksdb/ha_rocksdb.cc:4039
#5  0x0000000007b0337c in myrocks::ha_rocksdb::unlock_row (this=0x7fffbc883428) at /home/herman/rocks-mysql/8.0/storage/rocksdb/ha_rocksdb.cc:10463
#6  0x0000000005cdd1a0 in Sql_cmd_update::update_single_table (this=0x7fffbc80faf0, thd=0x7fffbc80b000)
    at /home/herman/rocks-mysql/8.0/sql/sql_update.cc:809
#7  0x0000000005ce0c47 in Sql_cmd_update::execute_inner (this=0x7fffbc80faf0, thd=0x7fffbc80b000)
    at /home/herman/rocks-mysql/8.0/sql/sql_update.cc:1614
#8  0x0000000005bf58be in Sql_cmd_dml::execute (this=0x7fffbc80faf0, thd=0x7fffbc80b000) at /home/herman/rocks-mysql/8.0/sql/sql_select.cc:738
#9  0x0000000005b51f1e in mysql_execute_command (thd=0x7fffbc80b000, first_level=true, last_timer=0x7ffff53b39b0)
    at /home/herman/rocks-mysql/8.0/sql/sql_parse.cc:4204
#10 0x0000000005b4b38d in mysql_parse (thd=0x7fffbc80b000, parser_state=0x7ffff53b4560, last_timer=0x7ffff53b39b0)
    at /home/herman/rocks-mysql/8.0/sql/sql_parse.cc:6202
#11 0x0000000005b46cf9 in dispatch_command (thd=0x7fffbc80b000, com_data=0x7ffff53b4d98, command=COM_QUERY)
    at /home/herman/rocks-mysql/8.0/sql/sql_parse.cc:2226
#12 0x0000000005b49ee1 in do_command (thd=0x7fffbc80b000) at /home/herman/rocks-mysql/8.0/sql/sql_parse.cc:1560
#13 0x0000000005dcaa21 in handle_connection (arg=0x7fffcea982e0)
    at /home/herman/rocks-mysql/8.0/sql/conn_handler/connection_handler_per_thread.cc:330
#14 0x00000000079f00fb in pfs_spawn_thread (arg=0x7ffff748a620) at /home/herman/rocks-mysql/8.0/storage/perfschema/pfs.cc:2884
#15 0x00007ffff7d5d20c in start_thread (arg=0x7ffff53b6700) at pthread_create.c:479
#16 0x00007ffff7f0581f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

However, performance with partial indexes might be very poor because we sort the primary keys, set the range lock, and then we need to re-read/re-sort the primary keys. We'd want to lock based on the rows returned by the partial index iterator.

Alternatively, we could skip supporting range locks for partial indexes and use only point locks, but going down this road means the compatibility matrix among different myrocks features like ttl, range locking, partial indexes, etc. gets complicated quickly.

cc: @lth

Reply:
This assertion happens even just regular SK on this table schema.

Reply:
Some further thoughts. I think we can either range lock the PK or range lock the SK here. Range locking the PK would still be correct since any insert/updates/scans would still need to lock the PK. We may end up locking more than needed, but I think it simplifies the implementation.

Since the PK and SK share the same prefixes and the PK is read to find the prefixes needed for the SK, we can range lock the prefixes on the PK in get_prefix_from_start(). m_pk_iterator.seek() is used to find the next prefixes, so we need to range lock the start_key to the returned prefix. We can range lock from endpoint (start_key, 0) to the endpoint (returned_prefix, +inf) so that we don't need to individually lock each gap between rows of the returned_prefix.

We would need to also implement the read/lock/re-read logic using the m_pk_iterator here. Given how the partial index iterator uses m_pk_iterator, it's not clear to me pushing m_iter_should_use_locking to the m_pk_iterator is sufficient because the range locks need to cover the whole range from one prefix to the next.

Reply:
Currently, partial indexes will call always Seek() on the secondary key range before reading from the PK, and that theoretically should acquire the correct range locks, and partial indexes should just work without much additional changes.

What we'll end up is that the locking PK iterator will ensure that no new prefixes show up in between consecutive prefixes, and the locking SK iterator will ensure that no new keys can be added within a prefix. 

That said, I think there are a few problems:
1. I don't think we're acquiring range locks if the iterator is invalid (in general, it doesn't seem like we acquire range locks up to the iterator end bound).
2. We also need to make sure we're acquiring point locks on the SK even if they're not materialized (it doesn't seem like that is done in this diff though).
3. In the partial index seek code, we always expect that we'd be able to read some rows out after getting the next prefix from the PK. However, with iterator refreshing, that might not be true anymore, so we probably need to add a while loop there.

There also may be some opportunity for optimization, since partial indexes currently uses a point lock on the prefix key to stand in as a range lock over the prefix (we do something similar for unique keys), so we probably don't have to lock twice. Also, it might be cleaner to lock only on PK instead as described above.

Reply:
Yes, Rdb_iterator_partial::seek() does always call Rdb_iterator_base::seek() to lock the range. We should get the correct range locking out of the partial index iterator's base iterator. The main downside is the ranges might be very large if the SK doesn't have any materialized prefixes, but that's a performance/concurrency issue, and not a correctness one.

I'd think we'd also need to set m_iterator_pk.m_iter_should_use_locking based on the partial index iterator's setting, which I think is not currently happening. And yes, I think locking only on the PK might be cleaner. We might be able to get away with not having to acquire SK point locks if we lock the whole prefix range in the PK before doing anything.

Reply:
> This assertion happens even with a regular SK (not partial index) in this table schema.

Confirm.  I've checked the cause:  The code attempts to count `m_row_lock_count`, and it counts range lock acquisition as one lock.  For rows that didn't match the WHERE, `unlock_row` decrements the count. If there is more than one row in the range, the counter can go below zero and fire an assert.

Reply:
Ok finally committed 9f29d708d47db4aaf4e015697dd088d2ee408b44  (pushed into this PR, too)  with a solution for this.